README.md
# Square Payment Module for Zen Cart

## About

The Square payment plugin module allows you to connect your Square account to your online store.
You can also do refunds for online purchases directly from within the order screen in your store Admin.

The payment gateway is operated by Square, Inc, which also offers POS card-swipe and chip readers to handle card-present transactions you can conduct from your phone or tablet in-store or mobile.

- Setup is quick: start accepting card payments in less than 5 minutes, with your existing Square account.
- No monthly fees and no setup fees.
- PCI Compliant. Customer never leaves your store!
- [Standard rates](https://developer.squareup.com/docs/payments-pricing#pricing-for-handling-online-and-in-app-payments) for Online payments are 2.9% + $0.30 per transaction for US and Canada businesses, and 2.5% for UK businesses. (Contact Square to negotiate.)
- Funds are deposited in your bank account in 1-2 business days.
- (If you are processing over $250,000 in annual sales, call (415) 805-9183 to connect with a US/Canada Square representative to discuss custom rates.)


## Countries Supported:
See the "online payments" section of this page for the types of payments Square currently supports in your country: https://developer.squareup.com/docs/payment-card-support-by-country#card-for-payment-online-or-mobile


## Compatibility

(When released ... ) Version 2.x of this module:
 - works with Zen Cart 1.5.6 and 1.5.7
 - requires PHP 7.1 to 7.4
 - works with Square API version 3.20200528.1 or newer

Version 1.x of this module:
 - is compatible with Zen Cart® versions 1.5.4 to 1.5.7
 - requires a version of PHP between 5.4-7.4
 - works with Square API Versions up to 3.20200528.1


## Requirements

1. You will need a Square account, already validated and connected with your bank. You may create a Square merchant account at [squareup.com](https://squareup.com)
2. You need to create an App inside your Square account. See below for instructions.
3. You must be using SSL on your website


## Creating an App inside your Square account
1. Login at https://developer.squareup.com/apps to view the apps you've connected to your account.
2. Click to create a New Application for your Zen Cart store to access. Give it a name, such as "Webstore", and click Create at the bottom of the page.
3. You may be prompted to allow a multi-location upgrade to your account's API configuration. Proceed.
4. You will now be given access to some credentials. (If you already have a custom app and are trying to re-use it, you may need to revoke its token and regenerate one.)
5. Click on the OAuth tab. In the "**Redirect URL**" field, enter `https://your_store_url.com/square_handler.php` and click Save at the bottom. (Ensure the URL you enter points to the correct directory (or subdirectory) on your server.)
6. Now click the "Show Secret" button, and copy the Application Secret. You will need it in your store Admin.
7. Also click back on the Credentials tab, and copy the Application ID. You will need it in your store Admin.
8. See "Admin module configuration" below.


## API Versions
It is best to configure your Square's API Version to match the API version this module is written for. To manually check for the API version the module is built for, see the square.php payment module file.

When switching API versions in your Square Dashboard, you can ignore the "Webhooks" switch because this module doesn't use Webhooks presently.


## Installation

### Installing Files To Your Zen Cart
Once you've downloaded the [Square Plugin](https://www.zen-cart.com/downloads.php?do=file&id=156) zip file, unzip it.

That will give you a folder with several sub-folders and files. Simply upload the files from `files_to_upload` into the corresponding folders on your own store:

`/square_handler.php` (to be clear, this goes in the same folder as your existing ipn_main_handler.php file)

`/includes/classes/vendors/square` (whole directory)

`/includes/modules/payment/square.php`

`/includes/modules/payment/square_support/square_admin_notification.php`

`/includes/modules/payment/square_support/ZenCartConnectCreatePaymentRequest.php`

`/includes/languages/english/modules/payment/square.php`

`/includes/modules/pages/checkout_payment/jscript_square.php` (this file may be needed in alternate checkout flows too, see Compatibility section below)

**Note: You should NOT copy the README.md, LICENSE or CHANGELOG.md files to your live server.**


###Upgrading

Upgrade instructions: 

1. **Delete** the server's `/includes/classes/vendors/square` directory and the `/includes/modules/payment/square_support` directory (You will replace these in a moment, but this deletes older obsolete files first). 

2. Then upload all the files from the `files_to_upload` directory **just as you would in the Installation steps above**.

3. If you had placed extra copies of the `jscript_square.php` file in other checkout flow locations, re-copy that again using the updated file. (common example would be for the One-Page-Checkout addon, described later in this document)

4. Login to https://squareup.com/dashboard/apps and click on My Apps. Click the Manage App button next to your listed App.
 - Make sure you are in Production (otherwise transactions aren't real!!!)
 - Scroll down to the API Version settings, and make sure your Production API Version is at least version 2020-03-25 or later ... more specifically (in case this document is outdated), best to be sure it matches the version listed in the top of the /includes/modules/payment/square.php file.

 
## Configure the Square Module in your store Admin

1. This module needs 2 configuration settings from your Square account, from when you created your App inside Square:

	(Login to https://squareup.com/dashboard/apps and click on My Apps.  Next, if there is one, click the Manage App button next to your listed App to obtain the configuration settings. Alternatively, click the "Go To Developer Portal" button, or visit https://developer.squareup.com/apps to get there) :
   * Application ID (under "Credentials" tab)
   * Application Secret (under "OAuth" tab)  (NOTE: While you are there, make sure your Redirect URL has been set as described earlier)

2. In your Zen Cart Admin, go to the Modules->Payments menu, and click on Square. Then click Install. 

3. You will see where to enter your Application ID and Application Secret. After pasting those values, click the Update button. This will update the screen.

4. Below the Edit/Remove buttons for the module,  you will see an "ALERT:  Access Token Not Set" message, along with a reminder about the Redirect URL for your site and a green button, "Click Here to Login and Authorize Your Account."  This will NOT work if you do not have the Redirect URL correctly set in your Square account.

5. Click the green button. It will take you to your Square account and ask you to login and authorize the app for your store. Click Authorize. (If you're already logged in, it may also just quickly open and close a window.) It may look like it never left or did anything, but if you refresh the page, that green button and alert will go away. (If the green button remains after reloading the page, then you probably didn't put your Redirect URL into your Square account's OAuth settings page correctly. Go back and do that, then click the green button again.)

6. Click Edit in the payment module settings, and find the Location ID drop-down. Choose one. If you have more than one choice, choose the location you want your Zen Cart store's payments to be associated with. Click Update.


Done. Your checkout pages will now show a credit card payment option, powered by Square.


## Handling Refunds, Captures, Voids
### Refunds
When viewing a transaction in your store Admin, if it was paid using Square within the last 120 days, you will be shown the option to Refund some or all of the payment. Simply enter the amount you wish to refund. Then check the box to confirm, and click Submit. You will see a status message confirming the refund has been issued.

NOTE: You SHOULD do all refunds via your store Admin. While you might be able to start a refund from your Square Dashboard, that will NOT automatically update your orders in Zen Cart to show the change in status.  But if you do the refund from within your Zen Cart Admin, the refund will show in both places, as well as in the customer's order-status-history, which is a more friendly experience for them.

### Captures
If you've configured the module to "authorize only" instead of automatically capturing each "purchase", then you will also see an option to Capture the previously-authorized transaction. CAPTURES MUST BE DONE WITHIN 6 DAYS, or they will be voided automatically, releasing the hold on those funds for the customer.

NOTE: Captures can ONLY be done via your store Admin. You will not see uncaptured (authorize-only) transactions in your Square Dashboard.

### Voids
If you've configured the module to "authorize only", then you will see a Void option. This can be used to cancel an Authorization and release the "hold" on those funds back to the customer. 

NOTE: Voids can ONLY be done via your store Admin. You will not see authorize-only transactions in your Square Dashboard.


## Automatic Token Refreshing via cron
This is optional. For low-traffic stores, this is recommended.

In the initial setup, you clicked that "green button" to activate the module by setting what's called a Refresh Token. This token will expire in 45 days if no transactions are made. (30 days expiration, plus 15 days grace period.) After the 30 days you will see the green "Get Access Token" button again in your store Admin, and will need to click it to reactivate the module as you did during initial setup.

With regular transaction activity the token will refresh itself, but this can come at the expense of a brief delay for random customers. To avoid the possibility of such delays, or expiration due to inactivity, we recommend that you set up a "cron job" on your server to automatically handle token refreshes without interfering with customer checkout activity.

To set up the cron job, go to the cron settings in your website hosting company's control panel, and create a new task to run every 3 days at 1am. The command for the task will be:

`php /home/user/your/path/to/wherever/is/square_handler.php`

Again, if your token expires, Square will no longer show up during checkout for your customers. But you can fix it manually by clicking the green button to "login and authorize your account" as described in the Configuration section above.


## Testing
To perform test transactions, Square prefers that you simply set up a $1 product in your store, and test with a real credit card. You can refund the test transactions from inside your store Admin without any extra service fees.

(This module *intentionally* does not support the use of the Square Sandbox, at Square's request. Use your store's primary credentials as described earlier, and test using real credit cards and real transactions.)


## Support
If you are receiving errors from Square or are having problems configuring settings in your Square Account, contact Square Support.

If you are running into difficulty configuring the module in your store, or are getting "unusual" errors during checkout, report your issue on the Zen Cart Support Site at https://www.zen-cart.com/forum.php (create a free account profile in order to post)


### Common errors
1. `SQ-NONCE-FAILURE`: This can be caused by template differences, or javascript/jQuery conflicts in your checkout pages. 

	The module expects the checkout page to have a <form> element with a `name="checkout_payment"` in it.

	If you've changed the checkout flow such that `tpl_checkout_payment_default.php` template file (or whatever files your custom checkout flow uses), no longer has a form with that name, you may need to update your template accordingly.
	
	Additionally, some custom templates forget to copy the `jscript_framework.php` file from `/includes/templates/template_default/jscript/jscript_framework.php` to `/includes/templates/YOUR_TEMPLATE_FOLDER/jscript/jscript_framework.php`. Without this file some other (non-Square-related) checkout functions may break the javascript required for the Square module to work, thus resulting in `SQ-NONCE-FAILURE` errors. This can often be identified by finding "Undefined variable zcJS" in the browser's console error logs in the split second before the page refreshes after clicking submit.


2. If you find the input fields for credit card numbers are flat with no text box to type into, this typically means you're missing the `jscript_square.php` file as described earlier in the Installation Instructions. 

	Or you've got a firewall or browser plugin blocking access to Square's javascript code. 
	
	Or you've got a plugin having jQuery conflicts.

3. You must always charge at least $1.00; transactions less than 1.00 will be rejected.

4. Cannot generate an Access Token.  
	If you re-used a pre-existing "application" in your Square Apps menu, go create a New Application. Some older applications may be based on an older API specification, or their tokens may not work without being regenerated. You could click the option to regenerate a token for the old application; however, anything you've previously associated with that application may now be broken. Creating a new application for Zen Cart is the simplest approach.

5. "The styling of the Square input fields is different from the rest of my payment input fields."  
	The module contains some CSS markup in the very bottom of the `/includes/modules/pages/checkout_payment/jscript_square.php` file, intended to create uniformity for the custom fields Square creates. For 99% of stores you will NOT need to change this; but in rare cases you may want to alter this CSS slightly to suit your needs. 


### Problems with OAuth Token Generation
Common causes of problems with this:

1. Didn't upload `square_handler.php`

2. Uploaded `square_handler.php` into the wrong directory. It belongs in the "root" of your store's folders. (For reference, `ipn_main_handler.php` also exists in this directory.)

3. Didn't put the **OAuth Redirect URL** into Square's settings, or mistyped it. 


Less common, but possible causes:

1. You've got URL-rewriting rules set up in .htaccess or nginx, which are disallowing `square_handler.php` to be reached, probably because you're redirecting those requests to some other URL/file. Fix: allow `square_handler.php` to be accessed directly. You could clone any rules for `ipn_main_handler.php` for `square_handler.php` as a starting point.

2. You're running on a test site using a fake/unreachable URL (and therefore Square's servers can't reach it). Fix by using a real live website.

3. You're running behind a firewall that's got aggressive restrictions to files. There's nothing special required for `square_handler.php`: It's just a simple PHP file, and will be accessed via port 443 ... just like any other file. There should be no rule against it.



## Compatibility With Various Zen Cart Plugins

### One-Page Checkout by lat9
For One-Page Checkout to work, you must **copy** the `jscript_square.php` file from `/includes/modules/pages/checkout_payment/` into the `/includes/modules/pages/checkout_one/` folder.



---

_Copyright (c) 2020 Zen Ventures, LLC. All Rights reserved._


CHANGELOG.md
# Square Payment Module for Zen Cart

Version 1:
 - Requires PHP 7.4, 7.3, 7.2, 7.1, 7.0, 5.6, 5.5, or 5.4
 - Requires Zen Cart v1.5.4 or v1.5.5 or v1.5.6 or v1.5.7


0.77 - June 23, 2017 - First Release - based on SquareConnect PHP SDK 2.2.0

0.80 - June 24, 2017 - Improvements to initial setup.

0.81 - June 28, 2017 - Change made to accommodate a Square SDK Locations API limitation.

0.82 - July 31, 2017 - Fix a small currency conversion error. (Not triggered unless store is multi-currency)

0.90 - Aug 14, 2017 - Fix problem with automatic token refresh; and updated to Square Connect SDK v2.2.1

0.91 - Dec 18, 2017 - Fix compatibility with OnePageCheckout plugin. Also updated to Square Connect SDK v2.5.1

0.92 - Dec 28, 2017 - Update auto-refresh of admin page when getting an Access Token
 	Now when a token is obtained after pressing the green button in the Admin, the page should auto-refresh within 5 seconds to indicate that the token was successfully applied. If it doesn't, simply refresh the page and the green button should go away.

0.93 - May 2018 - Compatibility updates. This module REQUIRES PHP 5.4 or newer. Preferably 5.6 or 7.1. Specifically: PHP 5.3 is NOT supported by this module.

0.94 - May 2018 - Simply updates the database storage configuration to allow for the newer longer transaction-id responses, since these are used to empower the in-Admin refund feature. 

0.95 - Nov 2018 - Fixed strict ['id'] error, Fixed currency conversion error if Square account currency was different from the Store's Default Currency. Also upgraded the SquareConnect SDK from 2.5.1 to 2.20180918.1.

0.96 - Dec 2018 - Fixed a jQuery error which could cause confusion on checkout payment page if checkout rules suddenly made the Square module unavailable. Also upgraded the SquareConnect SDK to 2.20181205.0.

0.97 - March 15 2019 - Updated to send headers forcing older 20181205 API version until code is refactored to use Square's new API endpoints introduced on 2019-03-13. Only files changed were the square_handler.php and the main square.php payment module file. (No change to the vendors folder.)

1.0 - March 15, 2020 - Updated to Square's 3.20200226.0 API spec, which uses new endpoints and improves OAuth token renewal process. (Will require a re-authorization of the module after upgrading to this version. It may not happen for up to 30 days.)
Also includes brief product/shipping details in transaction comments.

1.1 - April 8, 2020 - Updated to SquareConnect SDK version 3.20200325.0. Also fixed a token-refresh bug from v1.0

1.2 - June 21, 2020 - Fixed refund bug and capture bug in v1.1

1.3 - June 21, 2020 - Updated to SquareConnect SDK version 3.20200528.1
(Note: This is the last release of SquareConnect SDK that is compatible with PHP 5. Future versions will require PHP 7.1+, and will be integrated to the SquarePHP SDK (not SquareConnect). A version 2 of this module will be released to work with the new SDK, but this will also limit which Zen Cart versions the module is compatible with. If you are not using Zen Cart 1.5.6 or newer and PHP 7.1 or newer, now is the time to upgrade!)

1.4 - June 21, 2020 - Updated admin sidebar to show API/SDK version numbers for easy reference. (only changed square.php)

1.5 - June 22, 2020 - Added improvements to catching errors that occur during capture/void/refund. (only changed square.php)




---

###Upgrading

Upgrade instructions: 

1. Delete two directories on your server:
 - /includes/classes/vendors/square  
 - /includes/modules/payment/square_support

  (You will replace these directories in the next step, but this deletes older obsolete files first). 

2. Then upload all the files from the files_to_upload directory, putting them into the same directory on the server as you find them in this zip.

3. Consider updating the API version in your Square account to match the version for which this module is written: log into your Square account Dashboard, click on Apps, click on your app for Zen Cart, click on Production, and use the API dropdown to update it. Ignore the Webhooks option as this module doesn't use webhooks at this present moment.


template_extras-should-not-be-needed/tpl_checkout_payment_default.php
<?php
/**
 * Page Template
 *
 * Loaded automatically by index.php?main_page=checkout_payment.<br />
 * Displays the allowed payment modules, for selection by customer.
 *
 * @package templateSystem
 * @copyright Copyright 2003-2016 Zen Cart Development Team
 * @copyright Portions Copyright 2003 osCommerce
 * @license http://www.zen-cart.com/license/2_0.txt GNU Public License V2.0
 * @version $Id: Author: DrByte  Sat Oct 17 21:58:04 2015 -0400 Modified in v1.5.5 $
 */
?>
<?php echo $payment_modules->javascript_validation(); ?>
<div class="centerColumn" id="checkoutPayment">
<?php echo zen_draw_form('checkout_payment', zen_href_link(FILENAME_CHECKOUT_CONFIRMATION, '', 'SSL'), 'post'); ?>
<?php echo zen_draw_hidden_field('action', 'submit'); ?>

<h1 id="checkoutPaymentHeading"><?php echo HEADING_TITLE; ?></h1>

<?php if ($messageStack->size('redemptions') > 0) echo $messageStack->output('redemptions'); ?>
<?php if ($messageStack->size('checkout') > 0) echo $messageStack->output('checkout'); ?>
<?php if ($messageStack->size('checkout_payment') > 0) echo $messageStack->output('checkout_payment'); ?>

<?php // ** BEGIN PAYPAL EXPRESS CHECKOUT **
      if (!$payment_modules->in_special_checkout()) {
      // ** END PAYPAL EXPRESS CHECKOUT ** ?>
<h2 id="checkoutPaymentHeadingAddress"><?php echo TITLE_BILLING_ADDRESS; ?></h2>

<div id="checkoutBillto" class="floatingBox back">
<?php if (MAX_ADDRESS_BOOK_ENTRIES >= 2) { ?>
<div class="buttonRow forward"><?php echo '<a href="' . zen_href_link(FILENAME_CHECKOUT_PAYMENT_ADDRESS, '', 'SSL') . '">' . zen_image_button(BUTTON_IMAGE_CHANGE_ADDRESS, BUTTON_CHANGE_ADDRESS_ALT) . '</a>'; ?></div>
<?php } ?>
<address><?php echo zen_address_label($_SESSION['customer_id'], $_SESSION['billto'], true, ' ', '<br />'); ?></address>
</div>

<div class="floatingBox important forward"><?php echo TEXT_SELECTED_BILLING_DESTINATION; ?></div>
<br class="clearBoth" />
<?php // ** BEGIN PAYPAL EXPRESS CHECKOUT **
      }
      // ** END PAYPAL EXPRESS CHECKOUT ** ?>

<fieldset id="checkoutOrderTotals">
<legend id="checkoutPaymentHeadingTotal"><?php echo TEXT_YOUR_TOTAL; ?></legend>
<?php
  if (MODULE_ORDER_TOTAL_INSTALLED) {
    $order_totals = $order_total_modules->process();
?>
<?php $order_total_modules->output(); ?>
<?php
  }
?>
</fieldset>

<?php
  $selection =  $order_total_modules->credit_selection();
  if (sizeof($selection)>0) {
    for ($i=0, $n=sizeof($selection); $i<$n; $i++) {
      if ($_GET['credit_class_error_code'] == $selection[$i]['id']) {
?>
<div class="messageStackError"><?php echo zen_output_string_protected($_GET['credit_class_error']); ?></div>

<?php
      }
      for ($j=0, $n2=sizeof($selection[$i]['fields']); $j<$n2; $j++) {
?>
<fieldset>
<legend><?php echo $selection[$i]['module']; ?></legend>
<?php echo $selection[$i]['redeem_instructions']; ?>
<div class="gvBal larger"><?php echo $selection[$i]['checkbox']; ?></div>
<label class="inputLabel"<?php echo ($selection[$i]['fields'][$j]['tag']) ? ' for="'.$selection[$i]['fields'][$j]['tag'].'"': ''; ?>><?php echo $selection[$i]['fields'][$j]['title']; ?></label>
<?php echo $selection[$i]['fields'][$j]['field']; ?>
</fieldset>
<?php
      }
    }
?>

<?php
    }
?>

<?php // ** BEGIN PAYPAL EXPRESS CHECKOUT **
      if (!$payment_modules->in_special_checkout()) {
      // ** END PAYPAL EXPRESS CHECKOUT ** ?>
<fieldset class="payment">
<legend><?php echo TABLE_HEADING_PAYMENT_METHOD; ?></legend>

<?php
  if (SHOW_ACCEPTED_CREDIT_CARDS != '0') {
?>

<?php
    if (SHOW_ACCEPTED_CREDIT_CARDS == '1') {
      echo TEXT_ACCEPTED_CREDIT_CARDS . zen_get_cc_enabled();
    }
    if (SHOW_ACCEPTED_CREDIT_CARDS == '2') {
      echo TEXT_ACCEPTED_CREDIT_CARDS . zen_get_cc_enabled('IMAGE_');
    }
?>
<br class="clearBoth" />
<?php } ?>

<?php
  $selection = $payment_modules->selection();

  if (sizeof($selection) > 1) {
?>
<p class="important"><?php echo TEXT_SELECT_PAYMENT_METHOD; ?></p>
<?php
  } elseif (sizeof($selection) == 0) {
?>
<p class="important"><?php echo TEXT_NO_PAYMENT_OPTIONS_AVAILABLE; ?></p>

<?php
  }
?>

<?php
  $radio_buttons = 0;
  for ($i=0, $n=sizeof($selection); $i<$n; $i++) {
?>
<?php
    if (sizeof($selection) > 1) {
        if (empty($selection[$i]['noradio'])) {
 ?>
<?php echo zen_draw_radio_field('payment', $selection[$i]['id'], ($selection[$i]['id'] == $_SESSION['payment'] ? true : false), 'id="pmt-'.$selection[$i]['id'].'"'); ?>
<?php   } ?>
<?php
    } else {

?>
<?php echo zen_draw_hidden_field('payment', $selection[$i]['id'], 'id="pmt-'.$selection[$i]['id'].'"'); ?>
<?php
    }
?>
<label for="pmt-<?php echo $selection[$i]['id']; ?>" class="radioButtonLabel"><?php echo $selection[$i]['module']; ?></label>

<?php
    if (defined('MODULE_ORDER_TOTAL_COD_STATUS') && MODULE_ORDER_TOTAL_COD_STATUS == 'true' and $selection[$i]['id'] == 'cod') {
?>
<div class="alert"><?php echo TEXT_INFO_COD_FEES; ?></div>
<?php
    } else {
      // echo 'WRONG ' . $selection[$i]['id'];
?>
<?php
    }
?>
<br class="clearBoth" />

<?php
    if (isset($selection[$i]['error'])) {
?>
    <div><?php echo $selection[$i]['error']; ?></div>

<?php
    } elseif (isset($selection[$i]['fields']) && is_array($selection[$i]['fields'])) {
?>

<div class="ccinfo">
<?php
      for ($j=0, $n2=sizeof($selection[$i]['fields']); $j<$n2; $j++) {
?>
<label <?php echo (isset($selection[$i]['fields'][$j]['tag']) ? 'for="'.$selection[$i]['fields'][$j]['tag'] . '" ' : ''); ?>class="inputLabelPayment"><?php echo $selection[$i]['fields'][$j]['title']; ?></label><?php echo $selection[$i]['fields'][$j]['field']; ?>
<br class="clearBoth" />
<?php
      }
?>
</div>
<br class="clearBoth" />
<?php
    }
    $radio_buttons++;
?>
<br class="clearBoth" />
<?php
  }
?>

</fieldset>
<?php // ** BEGIN PAYPAL EXPRESS CHECKOUT **
      } else {
        ?><input type="hidden" name="payment" value="<?php echo $_SESSION['payment']; ?>" /><?php
      }
      // ** END PAYPAL EXPRESS CHECKOUT ** ?>
<fieldset>
<legend><?php echo TABLE_HEADING_COMMENTS; ?></legend>
<?php echo zen_draw_textarea_field('comments', '45', '3'); ?>
</fieldset>

<?php
  if (DISPLAY_CONDITIONS_ON_CHECKOUT == 'true') {
?>
<fieldset>
<legend><?php echo TABLE_HEADING_CONDITIONS; ?></legend>
<div><?php echo TEXT_CONDITIONS_DESCRIPTION;?></div>
<?php echo  zen_draw_checkbox_field('conditions', '1', false, 'id="conditions"');?>
<label class="checkboxLabel" for="conditions"><?php echo TEXT_CONDITIONS_CONFIRM; ?></label>
</fieldset>
<?php
  }
?>

<div class="buttonRow forward" id="paymentSubmit"><?php echo zen_image_submit(BUTTON_IMAGE_CONTINUE_CHECKOUT, BUTTON_CONTINUE_ALT, 'id="paymentSubmitButton" onclick="submitFunction('.zen_user_has_gv_account($_SESSION['customer_id']).','.$order->info['total'].')"'); ?></div>

<div class="buttonRow back"><?php echo TITLE_CONTINUE_CHECKOUT_PROCEDURE . '<br />' . TEXT_CONTINUE_CHECKOUT_PROCEDURE; ?></div>

</form>
</div>
